### PR TITLE
Add Gemma4 Assistant Model and Speculative Decoding

### PR DESCRIPTION
**Pull request by sachinprasadhs**
## Description of the change

# Add Gemma4 Assistant Model and Speculative Decoding

## Overview

This PR introduces two new components for accelerated Gemma4 inference — a
dedicated MTP assistant model (`Gemma4AssistantCausalLM`) and a generic
`SpeculativeSampler`  along with a bug fix in `Task` that caused
`task.weights.h5` to incorrectly include backbone weights.

---

## 1. `Gemma4AssistantCausalLM`

Adds `keras_hub.models.Gemma4AssistantCausalLM`, a small companion model
designed exclusively for speculative decoding with Gemma4 target models.

**How it works:**

- The assistant is a lightweight 4-layer Gemma4 model that shares the target
  model's KV cache via the `num_kv_shared_layers` mechanism in
  `Gemma4Backbone`. This means the assistant never computes its own key/value
  projections, it reads directly from the target model's cache.
- At each draft step the assistant receives the **target model's last hidden
  state** and the **last accepted token embedding**, which are concatenated and
  projected into the assistant's hidden dimension. This conditions the draft on
  what the target model "knows", making proposals significantly more accurate
  than a cold start.
- The output head uses a sparse **centroid-based ordered embedding**
  (`_apply_centroid_logits`) that routes predictions through `num_centroids`
  vocabulary clusters, keeping only the top-k active centroids per step. This
  reduces the effective output dimension from ~262K to ~4K tokens, greatly
  accelerating each draft step.
- Compatible with `from_preset()` for loading from HuggingFace-format
  checkpoints alongside the target model.

**Preset example:**

```python
import keras_hub

target_lm = keras_hub.models.Gemma4CausalLM.from_preset("gemma4_instruct_2b")
assistant_lm = keras_hub.models.Gemma4AssistantCausalLM.from_preset(
    "gemma4_instruct_2b_assistant"
)
output = target_lm.generate(
    "Tell me about speculative decoding.",
    assistant_model=assistant_lm,
    max_length=256,
)
```

---

## 2. `SpeculativeSampler`

Adds `keras_hub.samplers.SpeculativeSampler`, a backend-agnostic sampler that
implements speculative decoding
([Leviathan et al., 2022](https://arxiv.org/abs/2211.17192)) as a first-class
Keras Hub primitive.

**Algorithm — three phases per outer iteration:**

1. **Draft phase** — Runs `draft_next` K times autoregressively to produce K
   candidate tokens and their probabilities.
2. **Verify phase** — Calls `verify_next` once to score all K+1 positions in a
   single parallel target forward pass.
3. **Accept/reject phase** — Each draft token is accepted or rejected via
   rejection sampling (stochastic, when `base_sampler` is set) or greedy
   matching. A guaranteed bonus token is always appended, so the minimum
   throughput is 1 token/step with best-case K+1 tokens/step.

**Key design choices:**

- Works with any target and draft model that expose a `call_with_cache`
  interface — not specific to Gemma4.
- `verify_next` can return a 2-tuple `(logits, cache)` or a 3-tuple
  `(logits, hidden_states, cache)`. When hidden states are returned they seed
  the next draft cycle, improving draft accuracy for MTP-style assistants.
- Draft tokens are always selected greedily (argmax); stochasticity enters only
  through the `p(target)/q(draft)` acceptance ratio, consistent with the
  original algorithm.
- Supports both stateless (JAX) and stateful (PyTorch/TensorFlow) backends.

**Example:**

```python
sampler = keras_hub.samplers.SpeculativeSampler(
    num_speculative_tokens=5,
    base_sampler=keras_hub.samplers.TopPSampler(p=0.95),
)
```

---

## 3. Bug fix: `task.weights.h5` incorrectly included backbone weights

`save_task_weights()` and `load_task_weights()` both passed a **set of weight
IDs** to `objects_to_skip`, but the Keras saving API expects a **list of layer
objects**. Passing IDs caused the skip logic to be silently ignored, so every
`task.weights.h5` save included the full backbone weights.

**Fix:** Pass `list(self.backbone._flatten_layers())` directly.

**Impact:** Task-only weight files now correctly contain only the task head
weights. For example, with `DenseNet121ImageClassifier` the `task.weights.h5`
file drops from ~33 MB to ~3 MB (~90% reduction). The same improvement applies
to all `ImageClassifier` and `TextClassifier` variants.

---

## Other changes

- **`Gemma4Backbone`**: adds `num_kv_shared_layers` and
  `return_shared_kv_states` support to enable KV cache sharing with the
  assistant model.
- **`Gemma4CausalLM`**: adds `generate()` integration for speculative decoding
  via an `assistant_model` kwarg — handles draft/verify coordination, cache
  management, and hot/cold path compilation.
- **`convert_gemma4_assistant.py`**: new HuggingFace → Keras Hub weight
  converter for assistant checkpoints.
- **`preset_loader.py` / `preset_utils.py`**: extended to support loading
  `Gemma4AssistantCausalLM` presets alongside target model presets, plus a
  path traversal security fix in the TF GFile cache path (upstream commit
  `66c641c`).
- **`sampler.py` / `serialization.py`**: minor updates to register
  `SpeculativeSampler` in the sampler registry and public API.

## Checklist

- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have followed the Keras Hub [Model contribution guidelines](https://github.com/keras-team/keras-hub/blob/master/CONTRIBUTING_MODELS.md) in making these changes.
- [x] I have followed the Keras Hub [API design guidelines](https://github.com/keras-team/keras-hub/blob/master/API_DESIGN_GUIDE.md) in making these changes.
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).

### CODE REVIEW

File: keras_hub/src/models/gemma4/gemma4_assistant_causal_lm.py [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+        # active logit, preventing probability collapse onto active positions.
+        flat_hs = batch * seq
+        scatter_idx = ops.reshape(selected_canonical, (flat_hs, n_tokens))
+        flat_logits = ops.reshape(selected_logits, (flat_hs, n_tokens))
+        # Global min across all active logits in the batch (scalar tensor).
+        min_active = ops.min(flat_logits) - ops.cast(1.0, flat_logits.dtype)
+        output = (
+            ops.ones((flat_hs, vocab_size), dtype=hidden_states.dtype)
+            * min_active
+        )
```
[the remark this answers is not in the document]
Comment (sachinprasadhs) [reply keras_hub/src/models/gemma4/gemma4_assistant_causal_lm.py:222]: ops.full with a computed tensor fill value breaks the PyTorch backend (torch.full requires a scalar). Keeping the ops.ones * min_active broadcast — it's the correct cross-backend pattern here.

File: keras_hub/src/models/gemma4/gemma4_causal_lm.py [hunk 2]
```diff
@@ -605,9 +741,57 @@ def generate(
...
+                self.generate_function = None  # force recompile
+
+            self._assistant_model = assistant_model
+
+        outputs = super().generate(
             inputs,
             max_length=max_length,
             stop_token_ids=stop_token_ids,
             strip_prompt=strip_prompt,
         )
```
[the remark this answers is not in the document]
Comment (sachinprasadhs) [reply keras_hub/src/models/gemma4/gemma4_causal_lm.py]: Done

File: keras_hub/src/models/gemma4/gemma4_assistant_causal_lm_test.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+from keras_hub.src.models.gemma4.gemma4_assistant_causal_lm import (
+    Gemma4AssistantCausalLM,
+)
+from keras_hub.src.models.gemma4.gemma4_backbone import Gemma4Backbone
+from keras_hub.src.models.gemma4.gemma4_causal_lm import Gemma4CausalLM
+from keras_hub.src.tests.test_case import TestCase
+
+
+class Gemma4AssistantTest(TestCase, parameterized.TestCase):
```
[the remark this answers is not in the document]
Comment (sachinprasadhs) [reply keras_hub/src/models/gemma4/gemma4_assistant_causal_lm_test.py:15]: Done

File: keras_hub/src/models/gemma4/gemma4_assistant_causal_lm.py [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+        if target_kv_source_local_idx is not None:
+            shared_kv_local = target_cache[:, target_kv_source_local_idx, ...]
+        else:
+            shared_kv_local = target_cache[:, num_target - 2, ...]
+
+        layer_types = self.backbone.layer_types or []
+        shared_kv_map = {
+            i: (shared_kv_global if lt == "full_attention" else shared_kv_local)
+            for i, lt in enumerate(layer_types)
+        }
```
[the remark this answers is not in the document]
Comment (sachinprasadhs) [reply keras_hub/src/models/gemma4/gemma4_assistant_causal_lm.py:320]: `layer_types` will always be derived for assistant model and this check is not necessary.

File: keras_hub/src/models/gemma4/gemma4_assistant_causal_lm.py [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+        self.backbone_hidden_size = backbone_hidden_size
+        self.num_centroids = num_centroids
+        self.centroid_intermediate_top_k = centroid_intermediate_top_k
+        self.use_ordered_embeddings = use_ordered_embeddings
+        self.num_speculative_tokens = num_speculative_tokens
+        self.backbone = backbone
+
+        hidden_size = backbone.hidden_dim
+        vocabulary_size = backbone.token_embedding.input_dim
+        self._vocab_size_per_centroid = vocabulary_size // num_centroids
```
[the remark this answers is not in the document]
Comment (sachinprasadhs) [reply keras_hub/src/models/gemma4/gemma4_assistant_causal_lm.py:91]: Done.

File: keras_hub/src/models/gemma4/gemma4_backbone.py [hunk 6]
```diff
@@ -269,6 +275,7 @@ def __init__(
 
         self.vision_encoder = vision_encoder
         self.audio_encoder = audio_encoder
+        self.layer_types = layer_types
```
[the remark this answers is not in the document]
Comment (sachinprasadhs) [reply keras_hub/src/models/gemma4/gemma4_backbone.py:278]: This is not required, conversion script is updated.. for any future version of the model, it will store `layer_types` converted from HF.

File: keras_hub/src/samplers/speculative_sampler.py [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+            # p_probs: (batch, k, vocab)  — the K verification distributions
+            p_probs = target_probs[:, :k, :]
+
+            # Refresh the target cache in draft_cache so the next draft cycle
+            # uses the updated K/V entries. Supports 2-tuple and 3-tuple forms.
+            if (
+                has_draft_cache
+                and has_cache
+                and isinstance(draft_cache, tuple)
+                and len(draft_cache) == 2
```
Comment (divyashreepathihalli) [inline keras_hub/src/samplers/speculative_sampler.py:225]: this checks for a specific cache structure. How can we make this more generic to support different types of draft models that might have different cache structures?

File: keras_hub/src/models/gemma4/gemma4_backbone.py [hunk 8]
```diff
@@ -294,34 +301,62 @@ def __init__(
...
-                _type = "global" if _is_g else "local"
-                for k in range(len(_non_shared_types) - 1, -1, -1):
-                    if _non_shared_types[k] == _type:
-                        _kv_source[j] = k
-                        break
-        else:
-            _kv_source = {}
+            if _first_kv_shared == 0:
+                # Assistant mode: all layers share KV externally.
+                # Mirror HF's hardcoded logic: assign arbitrary indices
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/gemma4/gemma4_backbone.py:308]: Is this hardcoded logic robust? What if the target model has a different layer distribution? Does this assume a specific target model architecture?

File: keras_hub/src/models/gemma4/gemma4_assistant_causal_lm.py [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+@keras_hub_export("keras_hub.models.Gemma4AssistantCausalLM")
+class Gemma4AssistantCausalLM(CausalLM):
+    """An MTP assistant model for speculative decoding with Gemma4.
+
+    This model drafts candidate tokens to be verified by a larger target model.
+    It conditions on the target model's last hidden state and last token
+    embedding, and borrows the target model's KV cache via the KV-sharing
+    mechanism in `Gemma4Backbone`.
+
+    This model must NOT be used standalone. Each `call_with_cache()` step
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/gemma4/gemma4_assistant_causal_lm.py:18]: add a check in call or __call__ that raises a clear NotImplementedError or ValueError if used standalone,

**Comment by divyashreepathihalli** [comment]
Thanks for the PR Sachin! I just have a few comments.

File: keras_hub/src/samplers/speculative_sampler.py (hunk 7 above)
Comment (sachinprasadhs) [reply keras_hub/src/samplers/speculative_sampler.py:225]: Done, added an instance check, so that Gemma4 instance will take this path. For other future model path we can add the new logic to take different path.
If we try to make it generic now without any other models, it would be highly assumption based approach and this logic is flexible and has no regression impact for the above logic suggested.

File: keras_hub/src/models/gemma4/gemma4_assistant_causal_lm.py [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+    @_layers.setter
+    def _layers(self, value):
+        # Use object.__setattr__ to bypass __setattr__ → tracker recursion.
+        object.__setattr__(self, "_layers_storage", value)
+
+    def _flatten_layers(self, include_self=True, recursive=True):
+        """Delegate to super(); the ``_layers`` property already adds extras."""
+        return super()._flatten_layers(
+            include_self=include_self, recursive=recursive
+        )
```
[the remark this answers is not in the document]
Comment (sachinprasadhs) [reply keras_hub/src/models/gemma4/gemma4_assistant_causal_lm.py]: Done, cleaned up the code.

File: keras_hub/src/models/gemma4/gemma4_causal_lm.py [hunk 11]
```diff
@@ -605,9 +752,64 @@ def generate(
...
+            ):
+                # Reuse compiled speculative graph.
+                self.sampler = cached_spec_sampler
+                self.generate_function = cached_spec_fn
+            else:
+                # Compile a new speculative graph.
+                self.sampler = SpeculativeSampler(
+                    num_speculative_tokens=num_spec,
+                    base_sampler=spec_base_sampler,
+                )
```
[the remark this answers is not in the document]
Comment (sachinprasadhs) [reply keras_hub/src/models/gemma4/gemma4_causal_lm.py:796]: Temperature will be part of spec_base_sampler, still added as an explicit argument.

File: keras_hub/src/samplers/speculative_sampler.py [hunk 12]
```diff
@@ -0,0 +1,452 @@
...
+                has_draft_cache
+                and has_cache
+                and isinstance(draft_cache, tuple)
+                and len(draft_cache) == 3
+            ):
+                current_draft_cache = (
+                    current_draft_cache[0],
+                    updated_cache,
+                    current_draft_cache[2],
+                )
```
[the remark this answers is not in the document]
Comment (sachinprasadhs) [reply keras_hub/src/samplers/speculative_sampler.py:239]: Added instance check for Gemma4 model to take this path.

File: keras_hub/src/models/gemma4/gemma4_backbone.py (hunk 8 above)
Comment (sachinprasadhs) [reply keras_hub/src/models/gemma4/gemma4_backbone.py:308]: Gemma4 specific assistant model will not have it's own KV cache.. it will use target model's KV cache.
This logic will only be triggered for assistant model, where all KV is shared externally which satisfies `_first_kv_shared = num_layers - num_kv_shared_layers`

And else part is handling all the dynamic `_non_shared_types` dynamically calculating the exact source layer for every shared layer.

File: keras_hub/src/models/gemma4/gemma4_assistant_causal_lm.py (hunk 9 above)
Comment (sachinprasadhs) [reply keras_hub/src/models/gemma4/gemma4_assistant_causal_lm.py:18]: That's good idea, added call and generate_step as well with the NotImplementedError.

**Comment by divyashreepathihalli** [comment]
Thank you!! Has the numerics been verified?

**Comment by sachinprasadhs** [comment]
Yes
Screenshot 2026-05-15 at 2 14 20 AM

**Review by divyashreepathihalli — APPROVED** [review]

File: keras_hub/src/models/gemma4/gemma4_assistant_causal_lm.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (sachinprasadhs) [reply keras_hub/src/models/gemma4/gemma4_assistant_causal_lm.py]: Need to keep this, because pre_projection, post_projection, and centroids are used exclusively inside call_with_cache() during speculative decoding, they cannot be wired into the static inputs -> outputs functional graph and we need to override in this way.
